core/controllers/async_download.py

```
@@ -0,0 +1,150 @@
+    file_format: str,
+    funds: list[str] | None = None,
+    organisations: list[str] | None = None,
+    regions: list[str] | None = None,
+    rp_start: str | None = None,
+    rp_end: str | None = None,
+    outcome_categories: list[str] | None = None,
+):
+    """Download data, store file in S3 and send an email to the user with the download link."""
+
+    response = download(
+        file_format=file_format,
+        funds=funds,
+        organisations=organisations,
+        regions=regions,
+        rp_start=rp_start,
+        rp_end=rp_end,
+        outcome_categories=outcome_categories,
+    )
+
+    if response.status_code != 200:
+        current_app.logger.error(
+            "Error downloading data from data-store/download endpoint. Status code: {response_status}",
+            extra=dict(response_status=response.status_code),
+        )
+        return
+
+    file_format_str: str = get_file_format_from_extension(file_format)
+    file_size_str: str = get_human_readable_file_size(len(response.data))
+
+    # Upload the file to S3 and get presigned URL
+    file_obj = io.BytesIO(response.data)
+    bucket = Config.AWS_S3_BUCKET_FIND_DATA_FILES
+    current_datetime = datetime.now().strftime("%Y-%m-%d-%H:%M:%S")
+    file_name = f"fund-monitoring-data-{current_datetime}.{file_format}"
+
+    find_service_base_url = Config.FIND_SERVICE_BASE_URL
+    if find_service_base_url:
+        find_service_download_url = find_service_base_url + "/download"
+
+        upload_find_file(file=file_obj, bucket=bucket, object_name=file_name)
+        presigned_url = generate_find_presigned_url(bucket=bucket, object_name=file_name, expiry=86400)
+
+        # email the URL to the user
+        send_email_for_find_download(
+            email_address=email_address,
+            download_url=presigned_url,
+            find_service_url=find_service_download_url,
+            file_format=file_format_str,
+            file_size_str=file_size_str,
+        )
```

```suggestion
    find_service_download_url = Config.FIND_SERVICE_BASE_URL + "/download"

    upload_find_file(file=file_obj, bucket=bucket, object_name=file_name)
    presigned_url = generate_find_presigned_url(bucket=bucket, object_name=file_name, expiry=86400)

    # email the URL to the user
    send_email_for_find_download(
        email_address=email_address,
        download_url=presigned_url,
        find_service_url=find_service_download_url,
        file_format=file_format_str,
        file_size_str=file_size_str,
    )
```

I don't think we need the conditional here. If `FIND_SERVICE_BASE_URL` is None, it's probably better for us to just error rather than silently do nothing. That way it will get reported to sentry and we can fix it. If we do nothing, we won't find out until a user gets in contact to ask why they haven't received anything.

---

Changed back to as it was previously.